managed/services/backup/compatibility_service.go

```diff
@@ -0,0 +1,190 @@
+// Copyright (C) 2017 Percona LLC
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+package backup
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"gopkg.in/reform.v1"
+
+	"github.com/percona/pmm/managed/models"
+	"github.com/percona/pmm/managed/services/agents"
+)
+
+// CompatibilityService is responsible for checking software and artifacts compatibility during backup and restore.
+type CompatibilityService struct {
+	db *reform.DB
+	v  versioner
+	l  *logrus.Entry
+}
+
+// NewCompatibilityService creates new backups logic service.
+func NewCompatibilityService(db *reform.DB, v versioner) *CompatibilityService {
+	return &CompatibilityService{
+		l:  logrus.WithField("component", "management/backup/compatibility"),
+		db: db,
+		v:  v,
+	}
+}
+
+// checkSoftwareCompatibilityForService contains compatibility checking logic.
+func (s *CompatibilityService) checkCompatibility(serviceModel *models.Service, agentModel *models.Agent) (string, error) {
+	// Only MySQL compatibility checking implemented for now.
+	if serviceModel.ServiceType != models.MySQLServiceType {
+		return "", nil
+	}
+
+	softwares := []agents.Software{&agents.Mysqld{}, &agents.Xtrabackup{}, &agents.Xbcloud{}, &agents.Qpress{}}
+	svs, err := s.v.GetVersions(agentModel.AgentID, softwares)
+	if err != nil {
+		return "", err
+	}
+	if len(svs) != len(softwares) {
+		return "", errors.Wrapf(ErrComparisonImpossible, "response slice len %d != request len %d", len(svs), len(softwares))
+	}
+
+	svm := make(map[models.SoftwareName]string, len(softwares))
+	for i, software := range softwares {
+		name, err := convertSoftwareName(software)
+		if err != nil {
+			return "", err
+		}
+		if svs[i].Error != "" {
+			return "", errors.Wrapf(ErrComparisonImpossible, "failed to get software %s version: %s", name, svs[i].Error)
+		}
+
+		svm[name] = svs[i].Version
+	}
+
+	if err := mySQLSoftwaresInstalledAndCompatible(svm); err != nil {
+		return "", err
+	}
+
+	return svm[models.MysqldSoftwareName], nil
+}
+
+// findCompatibleServiceIDs looks for services compatible to artifact in given array of services' software versions.
+func (s *CompatibilityService) findCompatibleServiceIDs(artifactModel *models.Artifact, svs []*models.ServiceSoftwareVersions) []string {
```

Comment:
Using this naming `xxxModel` everywhere a little bit weird IMHO. But let's keep it for now and discuss it later on backend call.